File: sdk/python/agt.py
"""PCTR <-> Microsoft AGT bridge for Python.

AGT (https://github.com/microsoft/agent-governance-toolkit) is Python-first, so this
mirrors ``packages/pctr/src/agt.mjs`` for integrations that live on that side.

Types follow the toolkit's own definitions in
``agent-governance-typescript/src/types.ts`` — PolicyAction, PolicyDecisionResult,
AuditEntry, TrustScore, TrustTier, ExecutionRing, CascadeEvent, RingViolation.

The loop (docs/integration-guide.md Part 6):

1. AGT enforces pre-execution policy
2. PCTR observes the consequence, the route and the signed receipt
3. Trust is recomputed from that evidence
4. AGT consumes it and adjusts the next decision

AGT decides allow/deny. PCTR reports evidence and never a competing verdict.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, Dict, List, Mapping, Optional

__all__ = [
    "AGT_TIER_THRESHOLDS", "RING_BY_SEVERITY", "TrustScore", "trust_tier",
    "to_agt_trust_score", "from_agt_trust_score", "to_agt_score", "from_agt_score",
    "ring_for_severity", "is_spiffe_id", "parse_spiffe_id", "classify_action",
    "domain_for", "agt_claims", "to_trust_evidence", "to_mesh_attestation",
    "normalize_agt_event",
]

# ── Trust ────────────────────────────────────────────────────────────────────
# AGT's TrustScore is 0-1, banded into tiers. Thresholds are the toolkit's own
# defaults (agent-governance-typescript/src/trust.ts). A TTP score is already 0-1,
# so it maps across unscaled — that is the point of sharing a scale.
AGT_TIER_THRESHOLDS: Dict[str, float] = {
    "untrusted": 0.0, "provisional": 0.3, "trusted": 0.6, "verified": 0.85,
}


def _clamp01(value: Any) -> float:
    try:
        return max(0.0, min(1.0, float(value)))
    except (TypeError, ValueError):
        return 0.0


def trust_tier(score: Any, thresholds: Optional[Mapping[str, float]] = None) -> str:
    """Return AGT's TrustTier for a 0-1 score."""
    t = dict(AGT_TIER_THRESHOLDS if thresholds is None else thresholds)
    value = _clamp01(score)
    if value >= t["verified"]:
        return "Verified"
    if value >= t["trusted"]:
        return "Trusted"
    if value >= t["provisional"]:
        return "Provisional"
    return "Untrusted"


@dataclass
class TrustScore:
    """AGT's TrustScore: ``{ overall, dimensions, tier }``."""

    overall: float
    dimensions: Dict[str, float] = field(default_factory=dict)
    tier: str = ""

    def __post_init__(self) -> None:
        self.overall = _clamp01(self.overall)
        if not self.tier:
            self.tier = trust_tier(self.overall)

    def as_dict(self) -> Dict[str, Any]:
        return {"overall": self.overall, "dimensions": self.dimensions, "tier": self.tier}


def to_agt_trust_score(ttp_score: Any, dimensions: Optional[Mapping[str, float]] = None,
                       thresholds: Optional[Mapping[str, float]] = None) -> TrustScore:
    overall = _clamp01(ttp_score)
    return TrustScore(overall=overall, dimensions=dict(dimensions or {}),
                      tier=trust_tier(overall, thresholds))


def from_agt_trust_score(trust_score: Any) -> float:
    """Accept a TrustScore, a mapping, or a bare number."""
    if isinstance(trust_score, TrustScore):
        return trust_score.overall
    if isinstance(trust_score, Mapping):
        return _clamp01(trust_score.get("overall", 0))
    return _clamp01(trust_score)


# The 0-1000 integer scale from integration-guide.md 6.4. Upstream AGT does NOT use it;
# this is only for downstream components that explicitly expect that scale.
def to_agt_score(ttp_score: Any) -> int:
    return round(_clamp01(ttp_score) * 1000)


def from_agt_score(agt_score: Any) -> float:
    try:
        return _clamp01(float(agt_score) / 1000)
    except (TypeError, ValueError):
        return 0.0


# AGT's ExecutionRing: Ring0 is the most privileged. PCTR proposes a ring from what the
# action can cause; AGT's own actionRings configuration stays authoritative.
RING_BY_SEVERITY: Dict[str, int] = {"CRITICAL": 0, "HIGH": 1, "MEDIUM": 2, "LOW": 3}


def ring_for_severity(severity: Optional[str]) -> int:
    return RING_BY_SEVERITY.get(severity or "", 3)


# ── SPIFFE ───────────────────────────────────────────────────────────────────
_SPIFFE = re.compile(r"^spiffe://([^/]+)(/.*)?$")


def is_spiffe_id(agent_id: Any) -> bool:
    return bool(_SPIFFE.match(str(agent_id or "")))


def parse_spiffe_id(agent_id: Any) -> Optional[Dict[str, str]]:
    match = _SPIFFE.match(str(agent_id or ""))
    if not match:
        return None
    return {"trust_domain": match.group(1), "path": match.group(2) or "", "id": str(agent_id)}


# ── Consequences ─────────────────────────────────────────────────────────────
# Mirrors packages/pctr/src/consequences.mjs. The verb decides over the namespace:
# payments.status reads, payments.transfer moves money.
_READ_VERB = re.compile(
    r"(?:^|[./_:-])(get|list|read|status|search|query|fetch|describe|show|view|count|preview)$",
    re.IGNORECASE)

_RULES = [
    (re.compile(r"payment|transfer|payout|refund|charge|wire|treasury|stripe|ledger\.move|\bpay\b", re.I), "MONEY_MOVED"),
    (re.compile(r"delete|destroy|drop|purge|truncate|wipe|erase|remove_all|terminate", re.I), "DATA_DELETED"),
    (re.compile(r"secret|credential|token|apikey|api_key|private_key|vault\.read|password", re.I), "SECRET_EXPOSED"),
    (re.compile(r"contract|sign|execute_agreement|onchain|smart_contract", re.I), "CONTRACT_EXECUTED"),
    (re.compile(r"device|actuator|valve|door|robot|plc|scada", re.I), "PHYSICAL_CHANGED"),
    (re.compile(r"disable_user|deactivate|suspend_user|user\.update|identity|scim|role\.assign", re.I), "IDENTITY_CHANGED"),
    (re.compile(r"grant|permission|policy\.attach|share|acl|invite", re.I), "ACCESS_GRANTED"),
    (re.compile(r"deploy|release|rollback|migrate|prod|k8s\.apply|helm", re.I), "PRODUCTION_CHANGED"),
    (re.compile(r"terraform|infra|instance|bucket|dns|firewall|network|cluster", re.I), "INFRA_MODIFIED"),
    (re.compile(r"send|email|slack|sms|post_message|notify|publish|tweet", re.I), "MESSAGE_SENT"),
    (re.compile(r"write|create|update|insert|upsert|patch|put", re.I), "DATA_WRITTEN"),
    (re.compile(r"read|get|list|search|query|select|fetch", re.I), "DATA_READ"),
]

_CONSEQUENCES = {
    "MONEY_MOVED": ("money moved", "HIGH", False),
    "DATA_DELETED": ("data deleted", "CRITICAL", False),
    "IDENTITY_CHANGED": ("identity disabled or changed", "HIGH", True),
    "SECRET_EXPOSED": ("secret exposed", "CRITICAL", False),
    "PRODUCTION_CHANGED": ("production changed", "HIGH", True),
    "INFRA_MODIFIED": ("infrastructure modified", "HIGH", True),
    "ACCESS_GRANTED": ("access granted", "HIGH", True),
    "MESSAGE_SENT": ("external message sent", "MEDIUM", False),
    "CONTRACT_EXECUTED": ("contract executed", "CRITICAL", False),
    "PHYSICAL_CHANGED": ("physical system changed", "CRITICAL", False),
    "DATA_WRITTEN": ("data written", "MEDIUM", True),
    "DATA_READ": ("data read", "LOW", True),
    "NONE": ("no protected consequence", "LOW", True),
}

_SEVERITY_RANK = {"LOW": 0, "MEDIUM": 1, "HIGH": 2, "CRITICAL": 3}


def severity_rank(severity: Optional[str]) -> int:
    return _SEVERITY_RANK.get(severity or "", 0)


def classify_action(action: str, params: Optional[Mapping[str, Any]] = None) -> Dict[str, Any]:
    """What can this action cause? Scale and money raise the stakes."""
    hints = dict(params or {})
    consequence = hints.get("consequence")
    if consequence not in _CONSEQUENCES:
        consequence = None
    if consequence is None:
        consequence = "DATA_READ" if _READ_VERB.search(action or "") else None
    if consequence is None:
        for pattern, name in _RULES:
            if pattern.search(action or ""):
                consequence = name
                break
    consequence = consequence or "NONE"

    label, severity, reversible = _CONSEQUENCES[consequence]
    severity = hints.get("severity", severity)
    reversible = hints.get("reversible", reversible)

    records = hints.get("recordsAffected", hints.get("records_affected", 0)) or 0
    amount = hints.get("amount", 0) or 0
    if records >= 1000 or amount >= 5000:
        severity = severity if severity_rank(severity) >= 3 else "CRITICAL"
    elif records >= 100 or amount >= 500:
        severity = severity if severity_rank(severity) >= 2 else "HIGH"

    return {
        "action": action, "consequence": consequence, "label": label,
        "severity": severity, "reversible": reversible,
        "protected": consequence != "NONE" and severity_rank(severity) >= 1,
        "required_ring": ring_for_severity(severity),
    }


_DOMAIN_BY_CONSEQUENCE = {
    "MONEY_MOVED": "payments", "DATA_DELETED": "data-destruction", "SECRET_EXPOSED": "secrets",
    "PRODUCTION_CHANGED": "prod-change", "INFRA_MODIFIED": "infrastructure",
    "IDENTITY_CHANGED": "identity", "ACCESS_GRANTED": "access", "MESSAGE_SENT": "communications",
    "CONTRACT_EXECUTED": "contracts", "PHYSICAL_CHANGED": "physical",
    "DATA_WRITTEN": "data-write", "DATA_READ": "data-read",
}


def domain_for(consequence: Optional[str]) -> str:
    return _DOMAIN_BY_CONSEQUENCE.get(consequence or "", "general")


def _now() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


# ── OPA/Rego bridge (integration-guide.md 6.2) ───────────────────────────────
def agt_claims(*, ttp_score: float, action: Optional[str] = None,
               consequence: Optional[str] = None, severity: Optional[str] = None,
               reversible: Optional[bool] = None, domain: Optional[str] = None,
               issuer_count: int = 0, agents: Optional[List[str]] = None,
               route_id: Optional[str] = None, receipt: Optional[Mapping[str, Any]] = None,
               at: Optional[str] = None) -> Dict[str, Any]:
    """Build ``input.ttp`` for AGT's policy engine. Rego stays authoritative."""
    if action and not consequence:
        classified = classify_action(action)
        consequence = classified["consequence"]
        severity = severity or classified["severity"]
        reversible = classified["reversible"] if reversible is None else reversible

    agents = list(agents or [])
    score = _clamp01(ttp_score)
    return {
        "ttp": {
            "ttp_domain": domain or domain_for(consequence),
            "ttp_score": round(score, 4),
            "issuer_count": issuer_count,
            "ttp_agt_score": to_agt_score(score),            # legacy 0-1000 consumers
            "trust_score": to_agt_trust_score(score).as_dict(),  # AGT-native TrustScore
            "trust_tier": trust_tier(score),
            "required_ring": ring_for_severity(severity),
            "consequence": consequence,
            "severity": severity,
            "reversible": reversible,
            "route_id": route_id,
            "agents": agents,
            "spiffe_ids": [a for a in agents if is_spiffe_id(a)],
            "receipt_id": (receipt or {}).get("receiptId"),
            "receipt_hash": (receipt or {}).get("receiptHash"),
            "evaluated_at": at or _now(),
        }
    }


# ── Evidence back to AGT (step 2 of the loop) ────────────────────────────────
def to_trust_evidence(receipt: Optional[Mapping[str, Any]],
                      domain: Optional[str] = None) -> Optional[Dict[str, Any]]:
    if not receipt:
        return None
    verifier = receipt.get("verifier") or {}
    consequence = receipt.get("consequence") or {}
    allowed = verifier.get("decision") == "EXECUTION_ALLOWED"
    severity = consequence.get("severity", "LOW")
    executed = receipt.get("executed") or {}
    return {
        "type": "TTPBehavioralEvidence",
        "version": 1,
        "subject": receipt.get("requestedBy") or receipt.get("principal"),
        "domain": domain or domain_for(consequence.get("class")),
        "action": (receipt.get("requested") or {}).get("action"),
        "outcome": "allowed" if allowed else "denied",
        # A denial at a high-consequence action is the strongest behavioural signal.
        "weight": round((severity_rank(severity) + 1) / 4, 2),
        "failures": [f.get("code") for f in verifier.get("failures", [])],
        "receipt_id": receipt.get("receiptId"),
        "receipt_hash": receipt.get("receiptHash"),
        "observed_at": executed.get("executedAt") or receipt.get("issuedAt"),
    }


def to_mesh_attestation(receipt: Optional[Mapping[str, Any]], peer: Optional[str] = None,
                        trust_domain: Optional[str] = None) -> Optional[Dict[str, Any]]:
    """AgentMesh peer trust attestation (integration-guide.md 6.5)."""
    if not receipt:
        return None
    route = receipt.get("routeSelected") or {}
    principal = receipt.get("principal")
    parsed = parse_spiffe_id(principal) if principal else None
    score = _clamp01(route.get("effectiveTrust", 0))
    return {
        "type": "AgentMeshTrustAttestation",
        "version": 1,
        "subject": receipt.get("requestedBy") or principal,
        "peer": peer,
        "trust_domain": trust_domain or (parsed or {}).get("trust_domain"),
        "action": (receipt.get("requested") or {}).get("action"),
        "consequence": (receipt.get("consequence") or {}).get("class"),
        "outcome": (receipt.get("verifier") or {}).get("decision", "EXECUTION_DENIED"),
        "trust_score": to_agt_trust_score(score).as_dict(),
        "agt_trust_score": to_agt_score(score),
        "ttp_score": score,
        "evidence": {
            "receipt_id": receipt.get("receiptId"),
            "receipt_hash": receipt.get("receiptHash"),
            "signed_by": receipt.get("keyId"),
        },
        "issued_at": receipt.get("issuedAt"),
    }


# ── AGT events -> canonical PCTR events ──────────────────────────────────────
_POLICY_ACTION = {
    "allow": "EXECUTION_ALLOWED", "log": "EXECUTION_ALLOWED", "warn": "EXECUTION_ALLOWED",
    "deny": "EXECUTION_DENIED",
    "require_approval": "EXECUTION_DENIED",  # not denied forever, but not executable yet
}
_LEGACY_DECISION = {
    "allow": "EXECUTION_ALLOWED", "deny": "EXECUTION_DENIED", "review": "EXECUTION_DENIED",
}
# CascadeEvent actions all describe trust collapsing around an agent.
_CASCADE_TRUST = {
    "circuit_opened": 0.2, "agent_degraded": 0.3, "agent_quarantined": 0.0,
    "agent_killed": 0.0, "rollback_triggered": 0.4, "health_propagated": None,
}


def normalize_agt_event(event: Any) -> Optional[Dict[str, Any]]:
    """Map one AGT event onto a canonical PCTR event, or None if unrecognised.

    Returning None is a valid, common answer: PCTR never invents a security event
    from a shape it cannot read.
    """
    if not isinstance(event, Mapping):
        return None

    agent = (event.get("agentId") or event.get("agent_id") or event.get("sourceAgentId")
             or event.get("agent") or event.get("subject"))
    action = (event.get("action") or event.get("operation") or event.get("tool")
              or event.get("capability"))

    # CascadeEvent
    if event.get("eventId") and isinstance(event.get("affectedAgentIds"), list) and action in _CASCADE_TRUST:
        to = _CASCADE_TRUST[action]
        if to is None:
            return None  # health_propagated is telemetry, not a trust change
        return {"event": "TRUST_CHANGED", "subject": agent, "detail": {
            "to": to, "reason": event.get("reason") or f"AGT cascade containment: {action}",
            "cascade_action": action, "affected": event.get("affectedAgentIds"),
            "blast_radius": event.get("blastRadius")}}

    # RingViolation
    if event.get("requiredRing") is not None and event.get("agentRing") is not None:
        return {"event": "EXECUTION_DENIED", "subject": agent, "detail": {
            "action": event.get("action"),
            "reason": event.get("message") or
                      f"agent is Ring{event['agentRing']}; {event.get('action')} requires Ring{event['requiredRing']}",
            "agent_ring": event.get("agentRing"), "required_ring": event.get("requiredRing")}}

    # KillSwitchResult
    if event.get("killed") is True or event.get("terminated") is True:
        return {"event": "TRUST_CHANGED", "subject": agent, "detail": {
            "to": 0.0, "reason": event.get("reason") or "AGT kill switch engaged"}}

    # PolicyDecisionResult
    if isinstance(event.get("allowed"), bool) and event.get("action") in _POLICY_ACTION:
        policy_action = event["action"]
        pending = policy_action == "require_approval"
        approvers = event.get("approvers") or []
        return {
            "event": "EXECUTION_ALLOWED" if (event["allowed"] and not pending) else _POLICY_ACTION[policy_action],
            "subject": agent,
            "detail": {
                "action": event.get("operation") or event.get("tool"),
                "reason": event.get("reason") or (
                    f"AGT requires approval from: {', '.join(approvers) or 'an approver'}" if pending
                    else f"AGT policy {policy_action}"),
                "policy": event.get("policyName") or event.get("matchedRule"),
                "policy_action": policy_action,
                "approvers": approvers,
                "rate_limited": event.get("rateLimited", False),
            },
        }

    # AuditEntry — AGT hash-chains its audit log as PCTR chains receipts.
    if event.get("hash") and "previousHash" in event and isinstance(event.get("decision"), str):
        mapped = _LEGACY_DECISION.get(event["decision"])
        if not mapped:
            return None
        return {"event": mapped, "subject": agent, "detail": {
            "action": event.get("action"), "decision": event["decision"],
            "reason": f'AGT audit entry records decision "{event["decision"]}"',
            "audit_hash": event.get("hash"), "previous_hash": event.get("previousHash"),
            "skill": (event.get("skillAuditMetadata") or {}).get("skillName")}}

    # TrustVerificationResult
    if isinstance(event.get("verified"), bool) and event.get("trustScore") is not None:
        score = from_agt_trust_score(event["trustScore"])
        tier = event["trustScore"].get("tier") if isinstance(event["trustScore"], Mapping) else None
        return {"event": "TRUST_CHANGED", "subject": agent, "detail": {
            "to": score, "tier": tier or trust_tier(score),
            "reason": event.get("reason") or
                      f"AGT trust verification {'passed' if event['verified'] else 'failed'}"}}

    kind = str(event.get("type") or event.get("event") or event.get("eventType") or event.get("kind") or "").lower()

    if re.search(r"register|discover|onboard", kind) and agent:
        raw = event.get("trustScore")
        return {"event": "AGENT_DISCOVERED", "subject": agent, "detail": {
            "framework": "microsoft-agt",
            "trust": from_agt_trust_score(raw) if raw is not None else None,
            "status": event.get("status")}}

    if re.search(r"delegat|handoff|assign", kind):
        to = event.get("to") or event.get("target") or event.get("targetAgent") or event.get("delegateTo")
        return {"event": "AGENT_DELEGATED", "subject": agent, "detail": {"to": to}} if to else None

    if re.search(r"(action|tool|capability).*(request|invoke|propos|call)|invocation", kind):
        if not action:
            return None
        params = event.get("parameters") or event.get("params") or event.get("arguments") or event.get("input") or {}
        classified = classify_action(action, params)
        detail: Dict[str, Any] = {"action": action}
        detail.update(params if isinstance(params, Mapping) else {})
        detail.update({"consequence": classified["consequence"], "severity": classified["severity"],
                       "required_ring": classified["required_ring"]})
        return {"event": "ACTION_PROPOSED", "subject": agent, "detail": detail}

    if re.search(r"complete|finish|result", kind) and action:
        failed = bool(event.get("error") or event.get("failed"))
        return {"event": "EXECUTION_COMPLETED", "subject": agent,
                "detail": {"action": action, "status": "FAILED" if failed else "SUCCEEDED"}}

    return None

File: packages/pctr/src/agt.mjs
import { classifyAction, severityRank } from './consequences.mjs';

// MICROSOFT AGT (Agent Governance Toolkit) BRIDGE.
// Upstream: https://github.com/microsoft/agent-governance-toolkit
//
// Types here follow the toolkit's own definitions in
// agent-governance-typescript/src/types.ts — PolicyAction, PolicyDecisionResult,
// AuditEntry, TrustScore, TrustTier, ExecutionRing, CascadeEvent, RingViolation —
// not a guess at them.
//
// The loop is the one in docs/integration-guide.md Part 6: AGT enforces pre-execution
// policy, PCTR/TTP supplies the behavioural evidence, trust is recomputed, and AGT
// consumes the updated evidence on the next decision. PCTR reports *evidence*, never a
// competing verdict — building a parallel privilege model is explicitly out of scope.
//
//   1. AGT enforces pre-execution policy
//   2. PCTR observes the consequence, the route and the execution receipt
//   3. Trust is recomputed from that evidence
//   4. AGT consumes it and adjusts the next decision

// AGT's TrustScore is 0-1, banded into tiers. Thresholds are the toolkit's own
// defaults (agent-governance-typescript/src/trust.ts): untrusted 0.0, provisional 0.3,
// trusted 0.6, verified 0.85. PCTR's effective trust is already 0-1, so it maps across
// directly — no rescaling, which is the whole point of sharing a scale.
export const AGT_TIER_THRESHOLDS = { untrusted: 0.0, provisional: 0.3, trusted: 0.6, verified: 0.85 };

export function trustTier(score, thresholds = AGT_TIER_THRESHOLDS) {
  const s = clamp01(score);
  if (s >= thresholds.verified) return 'Verified';
  if (s >= thresholds.trusted) return 'Trusted';
  if (s >= thresholds.provisional) return 'Provisional';
  return 'Untrusted';
}

// Produces AGT's TrustScore { overall, dimensions, tier }.
export function toAgtTrustScore(ttpScore, { dimensions = {}, thresholds } = {}) {
  const overall = clamp01(ttpScore);
  return { overall, dimensions, tier: trustTier(overall, thresholds) };
}
export const fromAgtTrustScore = (trustScore) =>
  clamp01(typeof trustScore === 'number' ? trustScore : trustScore?.overall ?? 0);

// The 0-1000 integer scale from integration-guide.md 6.4. Upstream AGT does NOT use it —
// its TrustScore.overall is 0-1 — so this is only for downstream components that the
// guide says expect a 0-1000 scale. Prefer toAgtTrustScore() when talking to AGT itself.
export const toAgtScore = (ttpScore) => Math.round(clamp01(ttpScore) * 1000);
export const fromAgtScore = (agtScore) => clamp01((Number(agtScore) || 0) / 1000);

const clamp01 = (n) => Math.max(0, Math.min(1, Number(n) || 0));

// AGT's ExecutionRing (types.ts): Ring0 is the most privileged. PCTR proposes a ring
// from what the action can cause; AGT's own actionRings config stays authoritative.
export const RING_BY_SEVERITY = { CRITICAL: 0, HIGH: 1, MEDIUM: 2, LOW: 3 };
export const ringForSeverity = (severity) => RING_BY_SEVERITY[severity] ?? 3;

// SPIFFE/SVID compatibility (6.3): in SPIFFE-native deployments the SVID URI *is* the
// agent id, so it has to survive round-tripping unmodified.
const SPIFFE = /^spiffe:\/\/([^/]+)(\/.*)?$/;
export const isSpiffeId = (id) => SPIFFE.test(String(id ?? ''));
export function parseSpiffeId(id) {
  const match = SPIFFE.exec(String(id ?? ''));
  return match ? { trustDomain: match[1], path: match[2] ?? '', id: String(id) } : null;
}

// OPA/Rego bridge (6.2): claims exposed under `input.ttp` so AGT policy can evaluate
// current behavioural trust. Rego stays authoritative for allow/deny.
export function agtClaims({ route, preview, receipt, domain, issuerCount, at = new Date().toISOString() }) {
  const effectiveTrust = route?.effectiveTrust ?? route?.selected?.effectiveTrust ?? 0;
  const selected = route?.selected ?? route ?? null;
  const consequence = preview?.consequence ?? receipt?.consequence?.class ?? null;

  return {
    ttp: {
      // Claims named in the Rego example in integration-guide.md 6.2.
      ttp_domain: domain ?? domainFor(consequence),
      ttp_score: Number(effectiveTrust.toFixed?.(4) ?? effectiveTrust),
      issuer_count: issuerCount ?? selected?.trustStates?.length ?? 0,
      // PCTR's additions: what the action can cause, and over which path.
      ttp_agt_score: toAgtScore(effectiveTrust),          // legacy 0-1000 consumers
      trust_score: toAgtTrustScore(effectiveTrust),       // AGT-native TrustScore
      trust_tier: trustTier(effectiveTrust),
      required_ring: ringForSeverity(preview?.severity ?? receipt?.consequence?.severity),
      consequence,
      severity: preview?.severity ?? receipt?.consequence?.severity ?? null,
      reversible: preview?.reversible ?? receipt?.consequence?.reversible ?? null,
      route_id: selected?.routeId ?? receipt?.routeSelected?.routeId ?? null,
      agents: selected?.agents ?? receipt?.routeSelected?.agents ?? [],
      spiffe_ids: (selected?.agents ?? []).filter(isSpiffeId),
      receipt_id: receipt?.receiptId ?? null,
      receipt_hash: receipt?.receiptHash ?? null,
      evaluated_at: at
    }
  };
}

const DOMAIN_BY_CONSEQUENCE = {
  MONEY_MOVED: 'payments', DATA_DELETED: 'data-destruction', SECRET_EXPOSED: 'secrets',
  PRODUCTION_CHANGED: 'prod-change', INFRA_MODIFIED: 'infrastructure', IDENTITY_CHANGED: 'identity',
  ACCESS_GRANTED: 'access', MESSAGE_SENT: 'communications', CONTRACT_EXECUTED: 'contracts',
  PHYSICAL_CHANGED: 'physical', DATA_WRITTEN: 'data-write', DATA_READ: 'data-read'
};
export const domainFor = (consequence) => DOMAIN_BY_CONSEQUENCE[consequence] ?? 'general';

// AgentMesh peer trust attestation bridge (6.5): a PCTR receipt becomes a mesh
// attestation, carrying its identifiers so mesh telemetry stays cryptographically
// traceable back to the execution it describes.
export function toMeshAttestation(receipt, { peer, trustDomain } = {}) {
  if (!receipt) return null;
  return {
    type: 'AgentMeshTrustAttestation',
    version: 1,
    subject: receipt.requestedBy ?? receipt.principal ?? null,
    peer: peer ?? null,
    trustDomain: trustDomain ?? (parseSpiffeId(receipt.principal)?.trustDomain ?? null),
    action: receipt.requested?.action ?? null,
    consequence: receipt.consequence?.class ?? null,
    outcome: receipt.verifier?.decision ?? 'EXECUTION_DENIED',
    trustScore: toAgtTrustScore(receipt.routeSelected?.effectiveTrust ?? 0),
    agtTrustScore: toAgtScore(receipt.routeSelected?.effectiveTrust ?? 0),
    ttpScore: receipt.routeSelected?.effectiveTrust ?? 0,
    evidence: { receiptId: receipt.receiptId, receiptHash: receipt.receiptHash, signedBy: receipt.keyId ?? null },
    issuedAt: receipt.issuedAt
  };
}

// Step 2 of the loop: the behavioural evidence PCTR submits back after execution.
export function toTrustEvidence(receipt, { domain } = {}) {
  if (!receipt) return null;
  const allowed = receipt.verifier?.decision === 'EXECUTION_ALLOWED';
  const severity = receipt.consequence?.severity ?? 'LOW';
  return {
    type: 'TTPBehavioralEvidence',
    version: 1,
    subject: receipt.requestedBy ?? receipt.principal ?? null,
    domain: domain ?? domainFor(receipt.consequence?.class),
    action: receipt.requested?.action ?? null,
    outcome: allowed ? 'allowed' : 'denied',
    // A denial at a high-consequence action is the strongest behavioural signal there
    // is, so it is weighted by what the action could have caused.
    weight: Number(((severityRank(severity) + 1) / 4).toFixed(2)),
    failures: (receipt.verifier?.failures ?? []).map((f) => f.code),
    receiptId: receipt.receiptId,
    receiptHash: receipt.receiptHash,
    observedAt: receipt.executed?.executedAt ?? receipt.issuedAt
  };
}

// Normalizes AGT runtime events onto the canonical PCTR events, reading the toolkit's
// own shapes: PolicyDecisionResult, AuditEntry, CascadeEvent, RingViolation,
// KillSwitchResult and TrustVerificationResult. Anything unrecognised returns null —
// PCTR never invents a security event from a shape it cannot read.

// AGT PolicyAction -> what actually happens to the execution.
const POLICY_ACTION = {
  allow: 'EXECUTION_ALLOWED',
  log: 'EXECUTION_ALLOWED',
  warn: 'EXECUTION_ALLOWED',
  deny: 'EXECUTION_DENIED',
  require_approval: 'EXECUTION_DENIED'   // not denied forever, but not executable yet
};
// AGT LegacyPolicyDecision, used by AuditEntry and the client.
const LEGACY_DECISION = { allow: 'EXECUTION_ALLOWED', deny: 'EXECUTION_DENIED', review: 'EXECUTION_DENIED' };

// CascadeEvent actions all describe trust collapsing around an agent.
const CASCADE_TRUST = {
  circuit_opened: 0.2, agent_degraded: 0.3, agent_quarantined: 0, agent_killed: 0,
  rollback_triggered: 0.4, health_propagated: null
};

export function normalizeAgtEvent(e) {
  if (!e || typeof e !== 'object') return null;
  const agent = e.agentId ?? e.agent_id ?? e.sourceAgentId ?? e.agent ?? e.subject ?? null;
  const action = e.action ?? e.operation ?? e.tool ?? e.capability ?? null;

  // CascadeEvent: { eventId, timestamp, sourceAgentId, affectedAgentIds, action, reason, blastRadius }
  if (e.eventId && Array.isArray(e.affectedAgentIds) && action in CASCADE_TRUST) {
    const to = CASCADE_TRUST[action];
    if (to === null) return null;   // health_propagated is telemetry, not a trust change
    return {
      event: 'TRUST_CHANGED', subject: agent,
      detail: { to, reason: e.reason ?? `AGT cascade containment: ${action}`,
        cascadeAction: action, affected: e.affectedAgentIds, blastRadius: e.blastRadius ?? null }
    };
  }

  // RingViolation: { action, agentRing, requiredRing, message }
  if (e.requiredRing !== undefined && e.agentRing !== undefined) {
    return {
      event: 'EXECUTION_DENIED', subject: agent,
      detail: { action: e.action ?? null, reason: e.message ?? `agent is Ring${e.agentRing}; ${e.action} requires Ring${e.requiredRing}`,
        agentRing: e.agentRing, requiredRing: e.requiredRing }
    };
  }

  // KillSwitchResult: { agentId, action?, reason } with a killed/terminated flag.
  if (e.killed === true || e.terminated === true) {
    return { event: 'TRUST_CHANGED', subject: agent, detail: { to: 0, reason: e.reason ?? 'AGT kill switch engaged' } };
  }

  // PolicyDecisionResult: { allowed, action: PolicyAction, matchedRule?, policyName?, reason?, approvers, rateLimited, evaluatedAt }
  if (typeof e.allowed === 'boolean' && typeof e.action === 'string' && e.action in POLICY_ACTION) {
    const approvalPending = e.action === 'require_approval';
    return {
      event: e.allowed && !approvalPending ? 'EXECUTION_ALLOWED' : POLICY_ACTION[e.action],
      subject: agent,
      detail: {
        action: e.operation ?? e.tool ?? null,
        reason: e.reason ?? (approvalPending ? `AGT requires approval from: ${(e.approvers ?? []).join(', ') || 'an approver'}` : `AGT policy ${e.action}`),
        policy: e.policyName ?? e.matchedRule ?? null,
        policyAction: e.action,
        approvers: e.approvers ?? [],
        rateLimited: e.rateLimited ?? false
      }
    };
  }

  // AuditEntry: { timestamp, agentId, action, decision, hash, previousHash }
  // AGT hash-chains its audit log exactly as PCTR chains receipts, so the chain
  // identifiers are carried through rather than dropped.
  if (e.hash && e.previousHash !== undefined && typeof e.decision === 'string') {
    const mapped = LEGACY_DECISION[e.decision];
    if (!mapped) return null;
    return {
      event: mapped, subject: agent,
      detail: { action: e.action ?? null, decision: e.decision,
        reason: `AGT audit entry records decision "${e.decision}"`,
        auditHash: e.hash, previousHash: e.previousHash,
        skill: e.skillAuditMetadata?.skillName ?? null }
    };
  }

  // TrustVerificationResult: { verified, trustScore: { overall, dimensions, tier }, reason }
  if (typeof e.verified === 'boolean' && e.trustScore) {
    return {
      event: 'TRUST_CHANGED', subject: agent,
      detail: { to: fromAgtTrustScore(e.trustScore), tier: e.trustScore?.tier ?? trustTier(fromAgtTrustScore(e.trustScore)),
        reason: e.reason ?? `AGT trust verification ${e.verified ? 'passed' : 'failed'}` }
    };
  }

  // Discovery and delegation, from the toolkit's agent registry and mesh surfaces.
  const kind = String(e.type ?? e.event ?? e.eventType ?? e.kind ?? '').toLowerCase();
  if (/register|discover|onboard/.test(kind) && agent) {
    return { event: 'AGENT_DISCOVERED', subject: agent,
      detail: { framework: 'microsoft-agt', trust: e.trustScore != null ? fromAgtTrustScore(e.trustScore) : null,
        status: e.status ?? null } };
  }
  if (/delegat|handoff|assign/.test(kind)) {
    const to = e.to ?? e.target ?? e.targetAgent ?? e.delegateTo ?? null;
    return to ? { event: 'AGENT_DELEGATED', subject: agent, detail: { to } } : null;
  }
  // An invocation AGT is about to evaluate: PCTR supplies the consequence it omits.
  if (/(action|tool|capability).*(request|invoke|propos|call)|invocation/.test(kind)) {
    if (!action) return null;
    const params = e.parameters ?? e.params ?? e.arguments ?? e.input ?? {};
    const c = classifyAction(action, params);
    return { event: 'ACTION_PROPOSED', subject: agent,
      detail: { action, ...params, consequence: c.consequence, severity: c.severity, requiredRing: ringForSeverity(c.severity) } };
  }
  if (/complete|finish|result/.test(kind) && action) {
    return { event: 'EXECUTION_COMPLETED', subject: agent, detail: { action, status: e.error || e.failed ? 'FAILED' : 'SUCCEEDED' } };
  }
  return null;
}
